src/puppetlabs/pcp/client.clj

```diff
@@ -228,11 +245,15 @@
       ;; during shutdown. The websocket-connection atom could be reset to true between checking
       ;; connected? and dereferencing the atom to close the connection.
       (let [connection @websocket-connection]
-        (if (-connection-connected? connection)
-          (ws/close @connection)))
+        (when (-connection-connected? connection)
+          (.close @connection)))
       (catch java.util.concurrent.ExecutionException exception
         (log/debug exception (i18n/trs "exception while closing the connection; connection already closed"))))
-    (.stop websocket-client))
+    ;; Stop WebSocketClient on a new thread it is not running on
+    (try
+      @(future (LifeCycle/stop websocket-client))
```

Comment:
Not clear on why this needs to be in a future, especially when we are blocked waiting on the result.

Reply:
It needs to be on a different thread https://eclipse.dev/jetty/documentation/jetty-10/programming-guide/index.html#pg-client-websocket-stop - future is just the way to do that.

Reply:
My reading of that comment is when it is called within the context of the websocket-client's thread.  What context is `-close` called in?

Reply:
Assuming we do need it the example the provide does not wait on the result. I wonder if there is a possible deadlock here. I would remove the deref of the future, and put the try/catch block within the future body.